### ISSUE TITLE
BSOD: SYSTEM_SERVICE_EXCEPTION

### ISSUE THREAD

**Issue by inoperable** [openzfsonwindows/ZFSin#190]
This happens to me only when I try to mount an EXTERNAL drive (USB in this case), and this never happens when using INTERNAL drives (SATA/NVME). 

It happens **immediately** after issuing `zpool import` command, once the drive is found and zpool tries to mount the dataset it BSODs

```
nt!KeBugCheckEx:
fffff800`1dfe4e00 48894c2408      mov     qword ptr [rsp+8],rcx ss:0018:ffffcd08`b8334ad0=000000000000003b
2: kd> !analyze -v
*******************************************************************************
*                                                                             *
*                        Bugcheck Analysis                                    *
*                                                                             *
*******************************************************************************

SYSTEM_SERVICE_EXCEPTION (3b)
An exception happened while executing a system service routine.
Arguments:
Arg1: 0000000080000003, Exception code that caused the bugcheck
Arg2: fffff8001ded8013, Address of the instruction which caused the bugcheck
Arg3: ffffcd08b83353e0, Address of the context record for the exception that caused the bugcheck
Arg4: 0000000000000000, zero.

Debugging Details:
------------------

*** WARNING: Unable to verify timestamp for ZFSin.sys

KEY_VALUES_STRING: 1

    Key  : Analysis.CPU.Sec
    Value: 1

    Key  : Analysis.DebugAnalysisProvider.CPP
    Value: Create: 8007007e on RECTIFIER

    Key  : Analysis.DebugData
    Value: CreateObject

    Key  : Analysis.DebugModel
    Value: CreateObject

    Key  : Analysis.Elapsed.Sec
    Value: 1

    Key  : Analysis.Memory.CommitPeak.Mb
    Value: 70

    Key  : Analysis.System
    Value: CreateObject

    Key  : Dump.Attributes.InsufficientDumpfileSize
    Value: 1


DUMP_FILE_ATTRIBUTES: 0xc
  Insufficient Dumpfile Size
  Kernel Generated Triage Dump

BUGCHECK_CODE:  3b

BUGCHECK_P1: 80000003

BUGCHECK_P2: fffff8001ded8013

BUGCHECK_P3: ffffcd08b83353e0

BUGCHECK_P4: 0

CONTEXT:  ffffcd08b83353e0 -- (.cxr 0xffffcd08b83353e0)
rax=ffffcd08b8335e08 rbx=ffffcd08b8337018 rcx=0000000000000000
rdx=ffffcd08b8337018 rsi=ffffcd08b8336610 rdi=0000000000000000
rip=fffff8001ded8013 rsp=ffffcd08b8335de0 rbp=ffffcd08b83363f0
 r8=ffffcd08b8336610  r9=ffffcd08b8336450 r10=ffffcd08b8337860
r11=0000000000000000 r12=ffffcd08b8337860 r13=ffffcd08b8336dd8
r14=ffffcd08b8336450 r15=ffffcd08b8335ef0
iopl=0         nv up ei pl zr na po nc
cs=0010  ss=0018  ds=002b  es=002b  fs=0053  gs=002b             efl=00040246
nt!KeCheckStackAndTargetAddress+0x53:
fffff800`1ded8013 cc              int     3
Resetting default scope

BLACKBOXBSD: 1 (!blackboxbsd)


BLACKBOXNTFS: 1 (!blackboxntfs)


BLACKBOXWINLOGON: 1

CUSTOMER_CRASH_COUNT:  1

PROCESS_NAME:  zpool.exe

TRAP_FRAME:  ffffcd08b8336e80 -- (.trap 0xffffcd08b8336e80)
NOTE: The trap frame does not contain all registers.
Some register values may be zeroed or incorrect.
rax=0000000000000000 rbx=0000000000000000 rcx=ffffbd0a3236c000
rdx=ffffcd08b8337078 rsi=0000000000000000 rdi=0000000000000000
rip=0000000000000000 rsp=ffffcd08b8337018 rbp=0000000000000002
 r8=ffffcd08b83370a8  r9=ffffcd08b83370d0 r10=000000000000000e
r11=ffffcd08b8336f78 r12=0000000000000000 r13=0000000000000000
r14=0000000000000000 r15=0000000000000000
iopl=0         nv up ei pl zr na po nc
00000000`00000000 ??              ???
Resetting default scope

IP_IN_FREE_BLOCK: 0

STACK_TEXT:  
ffffcd08`b8337018 fffff800`2d822d26 : ffffbd0a`3236c000 00000000`00000016 00000000`00000001 ffffcd08`00000000 : 0x0
ffffcd08`b8337020 ffffbd0a`3236c000 : 00000000`00000016 00000000`00000001 ffffcd08`00000000 ffff9881`63dc0180 : ZFSin+0x1d2d26
ffffcd08`b8337028 00000000`00000016 : 00000000`00000001 ffffcd08`00000000 ffff9881`63dc0180 fffff800`1de0f09a : 0xffffbd0a`3236c000
ffffcd08`b8337030 00000000`00000001 : ffffcd08`00000000 ffff9881`63dc0180 fffff800`1de0f09a ffffbd0a`02013380 : 0x16
ffffcd08`b8337038 ffffcd08`00000000 : ffff9881`63dc0180 fffff800`1de0f09a ffffbd0a`02013380 fffff800`2d659a44 : 0x1
ffffcd08`b8337040 ffff9881`63dc0180 : fffff800`1de0f09a ffffbd0a`02013380 fffff800`2d659a44 ffffbd0a`00000000 : 0xffffcd08`00000000
ffffcd08`b8337048 fffff800`1de0f09a : ffffbd0a`02013380 fffff800`2d659a44 ffffbd0a`00000000 ffffbd0a`380f2000 : 0xffff9881`63dc0180
ffffcd08`b8337050 00000072`ffb80000 : 00000073`00000000 00000000`00000000 ffffbd0a`508f3080 fffff800`2d8235c4 : nt!KeSetEvent+0xea
ffffcd08`b83370e0 00000073`00000000 : 00000000`00000000 ffffbd0a`508f3080 fffff800`2d8235c4 ffffbd0a`3236c000 : 0x00000072`ffb80000
ffffcd08`b83370e8 00000000`00000000 : ffffbd0a`508f3080 fffff800`2d8235c4 ffffbd0a`3236c000 ffffbd0e`18c56bf0 : 0x00000073`00000000


STACK_COMMAND:  .trap 0xffffcd08b8336e80 ; kb

SYMBOL_NAME:  ZFSin+1d2d26

MODULE_NAME: ZFSin

IMAGE_NAME:  ZFSin.sys

BUCKET_ID_FUNC_OFFSET:  1d2d26

FAILURE_BUCKET_ID:  0x3B_80000003_ZFSin!unknown_function

OSPLATFORM_TYPE:  x64

OSNAME:  Windows 10

FAILURE_ID_HASH:  {af383e87-b7a7-7fa3-8fe5-8c201667161a}

Followup:     MachineOwner
---------
```

**Comment by lundman** [comment]
Could you do the `analyze -v` with the symbols for ZFSin loaded? The symbols are installed into C:\Program Files\OpenZFSonWindows\symbols so that it will lookup `ZFSin+0x1d2d26` into its proper name, file and line number.

**Comment by inoperable** [comment]
```
11: kd> !analyze -v
*******************************************************************************
*                                                                             *
*                        Bugcheck Analysis                                    *
*                                                                             *
*******************************************************************************

SYSTEM_THREAD_EXCEPTION_NOT_HANDLED_M (1000007e)
This is a very common bugcheck.  Usually the exception address pinpoints
the driver/function that caused the problem.  Always note this address
as well as the link date of the driver/image that contains this address.
Some common problems are exception code 0x80000003.  This means a hard
coded breakpoint or assertion was hit, but this system was booted
/NODEBUG.  This is not supposed to happen as developers should never have
hardcoded breakpoints in retail code, but ...
If this happens, make sure a debugger gets connected, and the
system is booted /DEBUG.  This will let us see why this breakpoint is
happening.
Arguments:
Arg1: ffffffff80000003, The exception code that was not handled
Arg2: fffff8008ab9c9d9, The address that the exception occurred at
Arg3: fffff504a73af2f8, Exception Record Address
Arg4: fffff504a73aeb30, Context Record Address

Debugging Details:
------------------


KEY_VALUES_STRING: 1

    Key  : Analysis.CPU.Sec
    Value: 1

    Key  : Analysis.DebugAnalysisProvider.CPP
    Value: Create: 8007007e on RECTIFIER

    Key  : Analysis.DebugData
    Value: CreateObject

    Key  : Analysis.DebugModel
    Value: CreateObject

    Key  : Analysis.Elapsed.Sec
    Value: 1

    Key  : Analysis.Memory.CommitPeak.Mb
    Value: 81

    Key  : Analysis.System
    Value: CreateObject

    Key  : Dump.Attributes.InsufficientDumpfileSize
    Value: 1


DUMP_FILE_ATTRIBUTES: 0xc
  Insufficient Dumpfile Size
  Kernel Generated Triage Dump

BUGCHECK_CODE:  7e

BUGCHECK_P1: ffffffff80000003

BUGCHECK_P2: fffff8008ab9c9d9

BUGCHECK_P3: fffff504a73af2f8

BUGCHECK_P4: fffff504a73aeb30

EXCEPTION_RECORD:  fffff504a73af2f8 -- (.exr 0xfffff504a73af2f8)
ExceptionAddress: fffff8008ab9c9d9 (ZFSin!kmem_error+0x0000000000000639)
   ExceptionCode: 80000003 (Break instruction exception)
  ExceptionFlags: 00000000
NumberParameters: 1
   Parameter[0]: 0000000000000000

CONTEXT:  fffff504a73aeb30 -- (.cxr 0xfffff504a73aeb30)
rax=0000000000000001 rbx=ffffb284700dc0c0 rcx=47a59d03065e0000
rdx=0000000000000000 rsi=fffff8008abb0ec0 rdi=ffffb2842fa3ea88
rip=fffff8008ab9c9d9 rsp=fffff504a73af530 rbp=0000000000000080
 r8=0000000000000001  r9=0000000000000000 r10=0000000000000010
r11=fffff504a73af060 r12=0000000000000142 r13=fffff8007c325440
r14=ffffb28423aaf080 r15=0000000000000000
iopl=0         nv up ei pl nz na pe nc
cs=0010  ss=0018  ds=002b  es=002b  fs=0053  gs=002b             efl=00040202
ZFSin!kmem_error+0x639:
fffff800`8ab9c9d9 cc              int     3
Resetting default scope

BLACKBOXBSD: 1 (!blackboxbsd)


BLACKBOXNTFS: 1 (!blackboxntfs)


BLACKBOXWINLOGON: 1

CUSTOMER_CRASH_COUNT:  1

PROCESS_NAME:  System

ERROR_CODE: (NTSTATUS) 0x80000003 - {EXCEPTION}  Breakpoint  A breakpoint has been reached.

EXCEPTION_CODE_STR:  80000003

EXCEPTION_PARAMETER1:  0000000000000000

EXCEPTION_STR:  0x80000003

STACK_TEXT:  
fffff504`a73af530 fffff800`8aba6261 : ffffb284`00000003 ffffb284`2b00f040 fffff800`8c0d2040 00000000`00000000 : ZFSin!kmem_error+0x639 [c:\src\zfsin\zfsin\spl\module\spl\spl-kmem.c @ 1036] 
fffff504`a73af5e0 fffff800`8adf8a56 : fffff800`8c0d2040 fffff800`00000008 00000000`00002a00 00000000`00000001 : ZFSin!zfs_kmem_free+0x1c1 [c:\src\zfsin\zfsin\spl\module\spl\spl-kmem.c @ 2772] 
fffff504`a73af640 fffff800`8adf9a7c : fffff800`8c0d2040 fffff800`8ae88950 fffff504`a73af6f0 fffff800`00000000 : ZFSin!zfs_freesid+0x76 [c:\src\zfsin\zfsin\zfs\module\zfs\zfs_vnops_windows_lib.c @ 1445] 
fffff504`a73af680 fffff800`8adf195b : ffffb284`327f2238 00000000`00000000 00000000`00000001 00000000`00000000 : ZFSin!zfs_set_security+0x42c [c:\src\zfsin\zfsin\zfs\module\zfs\zfs_vnops_windows_lib.c @ 1610] 
fffff504`a73af760 fffff800`8ae0a492 : ffffb284`327c14b0 00000000`00000000 ffffb284`3105e000 fffff504`0000002c : ZFSin!zfs_znode_getvnode+0x1eb [c:\src\zfsin\zfsin\zfs\module\zfs\zfs_vnops_windows.c @ 1471] 
fffff504`a73af7c0 fffff800`8adb5ad0 : ffffb284`3105e000 00000000`0002e1fe fffff504`a73af8f0 00000000`00000000 : ZFSin!zfs_zget_ext+0x702 [c:\src\zfsin\zfsin\zfs\module\zfs\zfs_znode.c @ 1536] 
fffff504`a73af8b0 fffff800`8abb1349 : ffffb284`3105e000 ffffb284`00000002 ffffb284`2fa3eb48 fffff504`a73afb38 : ZFSin!zfs_unlinked_drain_task+0x120 [c:\src\zfsin\zfsin\zfs\module\zfs\zfs_dir.c @ 533] 
fffff504`a73afad0 fffff800`7b8ebad5 : ffffb284`2fa3ea88 fffff800`8abb0ec0 ffffb284`2fa3ea88 000fa46f`b19bbfff : ZFSin!taskq_thread+0x489 [c:\src\zfsin\zfsin\spl\module\spl\spl-taskq.c @ 1610] 
fffff504`a73afb90 fffff800`7b9ec49a : ffffc501`710c0180 ffffb284`700dc0c0 fffff800`7b8eba80 ffff9907`00000003 : nt!PspSystemThreadStartup+0x55
fffff504`a73afbe0 00000000`00000000 : fffff504`a73b0000 fffff504`a73a9000 00000000`00000000 00000000`00000000 : nt!KiStartSystemThread+0x2a


FAULTING_SOURCE_LINE:  c:\src\zfsin\zfsin\spl\module\spl\spl-kmem.c

FAULTING_SOURCE_FILE:  c:\src\zfsin\zfsin\spl\module\spl\spl-kmem.c

FAULTING_SOURCE_LINE_NUMBER:  1036

FAULTING_SOURCE_CODE:  
No source found for 'c:\src\zfsin\zfsin\spl\module\spl\spl-kmem.c'


SYMBOL_NAME:  ZFSin!kmem_error+639

MODULE_NAME: ZFSin

IMAGE_NAME:  ZFSin.sys

STACK_COMMAND:  .cxr 0xfffff504a73aeb30 ; kb

BUCKET_ID_FUNC_OFFSET:  639

FAILURE_BUCKET_ID:  AV_ZFSin!kmem_error

OSPLATFORM_TYPE:  x64

OSNAME:  Windows 10

FAILURE_ID_HASH:  {b9238a07-d776-8a5c-ed8d-5ceecdcbdaf3}

Followup:     MachineOwner
---------
```

**Comment by inoperable** [comment]
Just realised that those are 2 different dumps. Sorry about this - I deleted the first initial one and few more happened afterwards. The second is a **different** one, and I'm pretty sure those aren't related.
I'll recreate the on BSOD on external USB again and repost the proper dump here.

**Comment by inoperable** [comment]
This is the proper one - I reattached, tried to import and zpool **BSODed** immediately. 
Ignore the previous one, it's not needed/unrelated delete it - no need to clutter things up.
I can put new issue together and put the proper dump/description in it, since the 2nd dump
happens under different conditions, no need for more clutter and confusion ;-) 
let me know if that's needed. 
```
2: kd> !analyze -v
*******************************************************************************
*                                                                             *
*                        Bugcheck Analysis                                    *
*                                                                             *
*******************************************************************************

SYSTEM_SERVICE_EXCEPTION (3b)
An exception happened while executing a system service routine.
Arguments:
Arg1: 0000000080000003, Exception code that caused the bugcheck
Arg2: fffff805114d8013, Address of the instruction which caused the bugcheck
Arg3: fffffb01448fd3e0, Address of the context record for the exception that caused the bugcheck
Arg4: 0000000000000000, zero.

Debugging Details:
------------------

*** WARNING: Unable to verify timestamp for ZFSin.sys

KEY_VALUES_STRING: 1

    Key  : Analysis.CPU.Sec
    Value: 1

    Key  : Analysis.DebugAnalysisProvider.CPP
    Value: Create: 8007007e on RECTIFIER

    Key  : Analysis.DebugData
    Value: CreateObject

    Key  : Analysis.DebugModel
    Value: CreateObject

    Key  : Analysis.Elapsed.Sec
    Value: 1

    Key  : Analysis.Memory.CommitPeak.Mb
    Value: 75

    Key  : Analysis.System
    Value: CreateObject

    Key  : Dump.Attributes.InsufficientDumpfileSize
    Value: 1


DUMP_FILE_ATTRIBUTES: 0xc
  Insufficient Dumpfile Size
  Kernel Generated Triage Dump

BUGCHECK_CODE:  3b

BUGCHECK_P1: 80000003

BUGCHECK_P2: fffff805114d8013

BUGCHECK_P3: fffffb01448fd3e0

BUGCHECK_P4: 0

CONTEXT:  fffffb01448fd3e0 -- (.cxr 0xfffffb01448fd3e0)
rax=fffffb01448fde08 rbx=fffffb01448ff018 rcx=0000000000000000
rdx=fffffb01448ff018 rsi=fffffb01448fe610 rdi=0000000000000000
rip=fffff805114d8013 rsp=fffffb01448fdde0 rbp=fffffb01448fe3f0
 r8=fffffb01448fe610  r9=fffffb01448fe450 r10=fffffb01448ff860
r11=0000000000000000 r12=fffffb01448ff860 r13=fffffb01448fedd8
r14=fffffb01448fe450 r15=fffffb01448fdef0
iopl=0         nv up ei pl zr na po nc
cs=0010  ss=0018  ds=002b  es=002b  fs=0053  gs=002b             efl=00240246
nt!KeCheckStackAndTargetAddress+0x53:
fffff805`114d8013 cc              int     3
Resetting default scope

BLACKBOXBSD: 1 (!blackboxbsd)


BLACKBOXNTFS: 1 (!blackboxntfs)


BLACKBOXPNP: 1 (!blackboxpnp)


BLACKBOXWINLOGON: 1

CUSTOMER_CRASH_COUNT:  1

PROCESS_NAME:  zpool.exe

TRAP_FRAME:  fffffb01448fee80 -- (.trap 0xfffffb01448fee80)
NOTE: The trap frame does not contain all registers.
Some register values may be zeroed or incorrect.
rax=0000000000000000 rbx=0000000000000000 rcx=ffffe507bac42000
rdx=fffffb01448ff078 rsi=0000000000000000 rdi=0000000000000000
rip=0000000000000000 rsp=fffffb01448ff018 rbp=0000000000000002
 r8=fffffb01448ff0a8  r9=fffffb01448ff0d0 r10=000000000000000e
r11=fffffb01448fef78 r12=0000000000000000 r13=0000000000000000
r14=0000000000000000 r15=0000000000000000
iopl=0         nv up ei pl zr na po nc
00000000`00000000 ??              ???
Resetting default scope

IP_IN_FREE_BLOCK: 0

STACK_TEXT:  
fffffb01`448ff018 fffff805`1d622d26 : ffffe507`bac42000 00000000`00000016 00000000`00000001 fffffb01`00000000 : 0x0
fffffb01`448ff020 fffff805`1d6235c4 : ffffe507`bac42000 ffffe507`b0211d70 00000000`00000000 fffff805`1d451019 : ZFSin!vdev_open+0x396 [c:\src\zfsin\zfsin\zfs\module\zfs\vdev.c @ 1623] 
fffffb01`448ff100 fffff805`1d64f419 : ffffe507`8efbc000 ffffe506`75017040 00000000`00000000 fffff805`1d451019 : ZFSin!vdev_open_children+0x74 [c:\src\zfsin\zfsin\zfs\module\zfs\vdev.c @ 1547] 
fffffb01`448ff150 fffff805`1d622d26 : ffffe507`8efbc000 fffffb01`448ff1f8 fffffb01`448ff228 fffffb01`448ff250 : ZFSin!vdev_root_open+0x69 [c:\src\zfsin\zfsin\zfs\module\zfs\vdev_root.c @ 99] 
fffffb01`448ff1a0 fffff805`1d5f746c : ffffe507`8efbc000 fffff805`0000007f fffff805`1d730140 00000000`00000001 : ZFSin!vdev_open+0x396 [c:\src\zfsin\zfsin\zfs\module\zfs\vdev.c @ 1623] 
fffffb01`448ff280 fffff805`1d5f8862 : ffffe507`b879a000 ffffe507`b9d9d928 fffffb01`448ff338 00000000`00000000 : ZFSin!spa_ld_open_vdevs+0xfc [c:\src\zfsin\zfsin\zfs\module\zfs\spa.c @ 2819] 
fffffb01`448ff2c0 fffff805`1d5f6e85 : ffffe507`b879a000 fffff805`00000000 ffffe506`00000000 ffffe507`b0211d70 : ZFSin!spa_ld_trusted_config+0x372 [c:\src\zfsin\zfsin\zfs\module\zfs\spa.c @ 3188] 
fffffb01`448ff350 fffff805`1d5f92bf : ffffe507`b879a000 fffff805`00000000 fffffb01`448ff3c4 00000000`00000000 : ZFSin!spa_ld_mos_with_trusted_config+0x45 [c:\src\zfsin\zfsin\zfs\module\zfs\spa.c @ 4028] 
fffffb01`448ff390 fffff805`1d5f8ca3 : ffffe507`b879a000 ffffe507`00000000 fffffb01`448ff458 fffff805`1d4fe0a5 : ZFSin!spa_load_impl+0x11f [c:\src\zfsin\zfsin\zfs\module\zfs\spa.c @ 4074] 
fffffb01`448ff410 fffff805`1d6014a6 : ffffe507`b879a000 fffff805`00000003 fffffb01`00000000 00000000`00000000 : ZFSin!spa_load+0x53 [c:\src\zfsin\zfsin\zfs\module\zfs\spa.c @ 2405] 
fffffb01`448ff470 fffff805`1d680aeb : ffffe507`b9d9dc88 00000000`0000049c 00000000`00000000 fffffb01`448ff550 : ZFSin!spa_tryimport+0x176 [c:\src\zfsin\zfsin\zfs\module\zfs\spa.c @ 5846] 
fffffb01`448ff520 fffff805`1d68903f : ffffe507`b8797000 fffff805`1d680a90 00000000`00000000 00000000`00000008 : ZFSin!zfs_ioc_pool_tryimport+0x5b [c:\src\zfsin\zfsin\zfs\module\zfs\zfs_ioctl.c @ 1782] 
fffffb01`448ff570 fffff805`1d6a98e9 : ffffe506`74ea8a70 ffffe506`71acc400 00000000`00000000 00000000`00000000 : ZFSin!zfsdev_ioctl+0x70f [c:\src\zfsin\zfsin\zfs\module\zfs\zfs_ioctl.c @ 7654] 
fffffb01`448ff660 fffff805`1d6a7317 : ffffe506`74ea8a70 ffffe506`71acc400 ffffe506`71acc4d0 00000000`00000000 : ZFSin!ioctlDispatcher+0x1f9 [c:\src\zfsin\zfsin\zfs\module\zfs\zfs_vnops_windows.c @ 3966] 
fffffb01`448ff6f0 fffff805`1140c4c5 : ffffe506`74ea8a70 ffffe506`71acc400 00000000`000011bc 00000000`00000001 : ZFSin!dispatcher+0xf7 [c:\src\zfsin\zfsin\zfs\module\zfs\zfs_vnops_windows.c @ 4763] 
fffffb01`448ff780 fffff805`117fd7e8 : fffffb01`448ffb00 ffffe506`71acc400 00000000`00000001 00000000`00000000 : nt!IofCallDriver+0x55
fffffb01`448ff7c0 fffff805`117ff875 : 00000000`00000000 fffffb01`448ffb00 00000000`00000000 fffffb01`448ffb00 : nt!IopSynchronousServiceTail+0x1a8
fffffb01`448ff860 fffff805`117fec06 : 0000001b`762f8f80 00000000`00000000 00000000`00000000 00000000`00000000 : nt!IopXxxControlFile+0x5e5
fffffb01`448ff9a0 fffff805`115f6775 : 00000000`00000000 00000000`00000000 00000000`00000000 0000001b`762f98c8 : nt!NtDeviceIoControlFile+0x56
fffffb01`448ffa10 00007ffb`5f70a744 : 00000000`00000000 00000000`00000000 00000000`00000000 00000000`00000000 : nt!KiSystemServiceCopyEnd+0x25
0000001b`762f9ed8 00000000`00000000 : 00000000`00000000 00000000`00000000 00000000`00000000 00000000`00000000 : 0x00007ffb`5f70a744


STACK_COMMAND:  .trap 0xfffffb01448fee80 ; kb

FAULTING_SOURCE_LINE:  c:\src\zfsin\zfsin\zfs\module\zfs\vdev.c

FAULTING_SOURCE_FILE:  c:\src\zfsin\zfsin\zfs\module\zfs\vdev.c

FAULTING_SOURCE_LINE_NUMBER:  1623

FAULTING_SOURCE_CODE:  
  1619: 		vdev_set_state(vd, B_TRUE, VDEV_STATE_OFFLINE, VDEV_AUX_NONE);
  1620: 		return (SET_ERROR(ENXIO));
  1621: 	}
  1622: 
> 1623: 	error = vd->vdev_ops->vdev_op_open(vd, &osize, &max_osize, &ashift);
  1624: 
  1625: 	/*
  1626: 	 * Reset the vdev_reopening flag so that we actually close
  1627: 	 * the vdev on error.
  1628: 	 */


SYMBOL_NAME:  ZFSin!vdev_open+396

MODULE_NAME: ZFSin

IMAGE_NAME:  ZFSin.sys

BUCKET_ID_FUNC_OFFSET:  396

FAILURE_BUCKET_ID:  0x3B_80000003_ZFSin!vdev_open

OSPLATFORM_TYPE:  x64

OSNAME:  Windows 10

FAILURE_ID_HASH:  {70da04e9-71e9-4cbe-401b-4d13239e1004}

Followup:     MachineOwner
---------
```

**Comment by lundman** [comment]
OK so this is an existing pool that you are trying to import. Does just `zpool import` also panic (without the name, to list importable pools).

It then tries to open the pool, and call vdev_open() on it, based on the type of vdev it calls the correct open function, and it is NULL for some reason. How can you have a vdev-type that isn't supported ? Most curious.

**Comment by lundman** [comment]
If you go to the `vdev_open()` stack, and issue `dt vd` it'd be interesting to see which `vd->vdev_ops` ops it's pointing to. The only ones with a NULL open, is vdev_indirect, but it shouldn't be calling open with indirect.  Or if it's easier, send me the minidump.

**Comment by inoperable** [comment]
I forget to mention that I once fiddled with this particular dataset a bit and added/removed a cache device and a spare to this dataset, then removed them since cache didn't had any impact on the performance and spare won't help without proper redundancy first.
Removal of one vdev is in the status message of the dataset itself.

`zpool status`, archlinux/kernel 5.4.0.1/zfs 0.8.2-1
```
  pool: drop
 state: ONLINE
status: Some supported features are not enabled on the pool. The pool can
	still be used, but some features are unavailable.
action: Enable all features using 'zpool upgrade'. Once this is done,
	the pool may no longer be accessible by software that does not support
	the features. See zpool-features(5) for details.
  scan: scrub repaired 0B in 0 days 15:24:00 with 0 errors on Wed Oct 30 14:50:59 2019
remove: Removal of vdev 1 copied 40K in 0h0m, completed on Tue Oct 15 15:34:17 2019
    144 memory used for removed device mappings
config:

	NAME          STATE     READ WRITE CKSUM
	drop          ONLINE       0     0     0
	  sde1        ONLINE       0     0     0
```

**Comment by inoperable** [comment]
zfs get all
```
AME  PROPERTY              VALUE                  SOURCE
drop  type                  filesystem             -
drop  creation              Fri Aug  2 13:18 2019  -
drop  used                  247G                   -
drop  available             195G                   -
drop  referenced            247G                   -
drop  compressratio         1.03x                  -
drop  mounted               yes                    -
drop  quota                 none                   default
drop  reservation           none                   default
drop  recordsize            128K                   default
drop  mountpoint            /Volumes/drop          local
drop  sharenfs              off                    default
drop  checksum              on                     default
drop  compression           lz4                    local
drop  atime                 off                    local
drop  devices               on                     default
drop  exec                  on                     default
drop  setuid                on                     default
drop  readonly              off                    default
drop  zoned                 off                    default
drop  snapdir               hidden                 default
drop  aclinherit            restricted             default
drop  createtxg             1                      -
drop  canmount              on                     default
drop  xattr                 on                     default
drop  copies                1                      default
drop  version               5                      -
drop  utf8only              off                    -
drop  normalization         none                   -
drop  casesensitivity       mixed                  -
drop  vscan                 off                    default
drop  nbmand                off                    default
drop  sharesmb              off                    default
drop  refquota              none                   default
drop  refreservation        none                   default
drop  guid                  994987371739487771     -
drop  primarycache          all                    default
drop  secondarycache        all                    default
drop  usedbysnapshots       0B                     -
drop  usedbydataset         247G                   -
drop  usedbychildren        2.81M                  -
drop  usedbyrefreservation  0B                     -
drop  logbias               latency                default
drop  objsetid              54                     -
drop  dedup                 off                    default
drop  mlslabel              none                   default
drop  sync                  standard               default
drop  dnodesize             legacy                 default
drop  refcompressratio      1.03x                  -
drop  written               247G                   -
drop  logicalused           256G                   -
drop  logicalreferenced     256G                   -
drop  volmode               default                default
drop  filesystem_limit      none                   default
drop  snapshot_limit        none                   default
drop  filesystem_count      none                   default
drop  snapshot_count        none                   default
drop  snapdev               hidden                 default
drop  acltype               posixacl               local
drop  context               none                   default
drop  fscontext             none                   default
drop  defcontext            none                   default
drop  rootcontext           none                   default
drop  relatime              on                     local
drop  redundant_metadata    all                    default
drop  overlay               off                    default
drop  encryption            off                    default
drop  keylocation           none                   default
drop  keyformat             none                   default
drop  pbkdf2iters           0                      default
drop  special_small_blocks  0                      default
```

**Comment by lundman** [comment]
Ok, so `Removal of vdev 1 copied 40K in 0h0m` probably means that `vdev_indirect.c` is in fact active, and there is a bug in the Windows `vdev` (missing commit maybe) making it try to call `vdev_indirect_open()` which doesn't exist.

My `vdev.c` is about 10 or so commits behind, so it is entirely likely.

**Comment by inoperable** [comment]
Let me check out where vdev_ops points to, new day, new BSOD :-)

**Comment by lundman** [comment]
Heh:

ZOL:
```
vdev_ops_t vdev_indirect_ops = {
	.vdev_op_open = vdev_indirect_open,
	.vdev_op_close = vdev_indirect_close,
	.vdev_op_asize = vdev_default_asize,
```

ZFSin:
```
vdev_ops_t vdev_indirect_ops = {
	NULL,//vdev_indirect_open,
	NULL,//vdev_indirect_close,
	NULL,//vdev_default_asize,
```

Oh dear.

**Comment by inoperable** [comment]
```
ZFSin!vdev_ops
   +0x000 vdev_op_open     : Ptr64     int 
   +0x008 vdev_op_close    : Ptr64     void 
   +0x010 vdev_op_asize    : Ptr64     unsigned int64 
   +0x018 vdev_op_io_start : Ptr64     void 
   +0x020 vdev_op_io_done  : Ptr64     void 
   +0x028 vdev_op_state_change : Ptr64     void 
   +0x030 vdev_op_need_resilver : Ptr64     boolean_t 
   +0x038 vdev_op_hold     : Ptr64     void 
   +0x040 vdev_op_rele     : Ptr64     void 
   +0x048 vdev_op_remap    : Ptr64     void 
   +0x050 vdev_op_xlate    : Ptr64     void 
   +0x058 vdev_op_type     : [16] Char
   +0x068 vdev_op_leaf     : boolean_t
```
yeah,  `+0x008 vdev_op_close    : Ptr64     void `

**Comment by inoperable** [comment]
```
kd> dx -r1 (*((ZFSin!vdev_ops *)0xfffff801243bb0a0))
(*((ZFSin!vdev_ops *)0xfffff801243bb0a0)) [Type: vdev_ops]
    [+0x000] vdev_op_open     : Unable to read memory at Address 0xfffff801243bb0a0
    [+0x008] vdev_op_close    : Unable to read memory at Address 0xfffff801243bb0a8
    [+0x010] vdev_op_asize    : Unable to read memory at Address 0xfffff801243bb0b0
    [+0x018] vdev_op_io_start : Unable to read memory at Address 0xfffff801243bb0b8
    [+0x020] vdev_op_io_done  : Unable to read memory at Address 0xfffff801243bb0c0
    [+0x028] vdev_op_state_change : Unable to read memory at Address 0xfffff801243bb0c8
    [+0x030] vdev_op_need_resilver : Unable to read memory at Address 0xfffff801243bb0d0
    [+0x038] vdev_op_hold     : Unable to read memory at Address 0xfffff801243bb0d8
    [+0x040] vdev_op_rele     : Unable to read memory at Address 0xfffff801243bb0e0
    [+0x048] vdev_op_remap    : Unable to read memory at Address 0xfffff801243bb0e8
    [+0x050] vdev_op_xlate    : Unable to read memory at Address 0xfffff801243bb0f0
    [+0x058] vdev_op_type     : "--- memory read error at address 0xfffff801`243bb0f8 ---" [Type: char [16]]
    [+0x068] vdev_op_leaf     : Unable to read memory at Address 0xfffff801243bb108
2: kd> dx -r1 (*((ZFSin!vdev_ops *)0xfffff801243bb600))
(*((ZFSin!vdev_ops *)0xfffff801243bb600)) [Type: vdev_ops]
    [+0x000] vdev_op_open     : Unable to read memory at Address 0xfffff801243bb600
    [+0x008] vdev_op_close    : Unable to read memory at Address 0xfffff801243bb608
    [+0x010] vdev_op_asize    : Unable to read memory at Address 0xfffff801243bb610
    [+0x018] vdev_op_io_start : Unable to read memory at Address 0xfffff801243bb618
    [+0x020] vdev_op_io_done  : Unable to read memory at Address 0xfffff801243bb620
    [+0x028] vdev_op_state_change : Unable to read memory at Address 0xfffff801243bb628
    [+0x030] vdev_op_need_resilver : Unable to read memory at Address 0xfffff801243bb630
    [+0x038] vdev_op_hold     : Unable to read memory at Address 0xfffff801243bb638
    [+0x040] vdev_op_rele     : Unable to read memory at Address 0xfffff801243bb640
    [+0x048] vdev_op_remap    : Unable to read memory at Address 0xfffff801243bb648
    [+0x050] vdev_op_xlate    : Unable to read memory at Address 0xfffff801243bb650
    [+0x058] vdev_op_type     : "--- memory read error at address 0xfffff801`243bb658 ---" [Type: char [16]]
    [+0x068] vdev_op_leaf     : Unable to read memory at Address 0xfffff801243bb668

2: kd> dx -r1 (*((ZFSin!vdev_ops *)0xfffff801243bab50))
(*((ZFSin!vdev_ops *)0xfffff801243bab50)) [Type: vdev_ops]
    [+0x000] vdev_op_open     : Unable to read memory at Address 0xfffff801243bab50
    [+0x008] vdev_op_close    : Unable to read memory at Address 0xfffff801243bab58
    [+0x010] vdev_op_asize    : Unable to read memory at Address 0xfffff801243bab60
    [+0x018] vdev_op_io_start : Unable to read memory at Address 0xfffff801243bab68
    [+0x020] vdev_op_io_done  : Unable to read memory at Address 0xfffff801243bab70
    [+0x028] vdev_op_state_change : Unable to read memory at Address 0xfffff801243bab78
    [+0x030] vdev_op_need_resilver : Unable to read memory at Address 0xfffff801243bab80
    [+0x038] vdev_op_hold     : Unable to read memory at Address 0xfffff801243bab88
    [+0x040] vdev_op_rele     : Unable to read memory at Address 0xfffff801243bab90
    [+0x048] vdev_op_remap    : Unable to read memory at Address 0xfffff801243bab98
    [+0x050] vdev_op_xlate    : Unable to read memory at Address 0xfffff801243baba0
    [+0x058] vdev_op_type     : "--- memory read error at address 0xfffff801`243baba8 ---" [Type: char [16]]
    [+0x068] vdev_op_leaf     : Unable to read memory at Address 0xfffff801243babb8
```

**Comment by lundman** [comment]
Yes, as you can see, I set them to NULL in Windows, should be trivial to fix, I'll have a new build once this unrelated code works.

**Comment by lundman** [comment]
ok, you can test that with 1124
https://openzfsonosx.org/wiki/Windows_builds

and as a head's up, this has new IO dispatch code, that skips the taskq. Works well here, but could be you have a corner case, so keep that in mind. But might be a little faster too.

**Comment by inoperable** [comment]
Running still the last release, on waking up from idle in Windows, my dataset crapped out with I/O Error. Rebooted to Linux and I run into trouble.

It might be a HW error - since the drive is OLD, 35xxx hours power-on, a hdd from 2008 - but it might be something else too. 

```
─Nov 22 08:34:02 [kernel] [  269.774772] WARNING: can't open objset 54, error 5                                                        │
┌Nov 22 08:34:02 [kernel] [  269.774772] WARNING: can't open objset 54, error 5                                                        │
│Nov 22 08:34:02 [zed] eid=40 class=data pool_guid=0x47641B044A0A279B                                                                  │
│Nov 22 08:34:02 [zed] eid=41 class=checksum pool_guid=0x47641B044A0A279B vdev_path=/dev/sdd2                                          │
│Nov 22 08:34:02 [zed] eid=42 class=checksum pool_guid=0x47641B044A0A279B vdev_path=/dev/sdd2                                          │
│Nov 22 08:34:02 [zed] eid=43 class=checksum pool_guid=0x47641B044A0A279B vdev_path=/dev/sdd2                                          │
│Nov 22 08:34:02 [zed] eid=44 class=checksum pool_guid=0x47641B044A0A279B vdev_path=/dev/sdd2                                          │
│Nov 22 08:34:02 [zed] eid=45 class=data pool_guid=0x47641B044A0A279B                                                                  │
│Nov 22 08:34:02 [zed] eid=46 class=checksum pool_guid=0x47641B044A0A279B vdev_path=/dev/sdd2                                          │
│Nov 22 08:34:03 [zed] eid=47 class=checksum pool_guid=0x47641B044A0A279B vdev_path=/dev/sdd2                                          │
│Nov 22 08:34:03 [zed] eid=48 class=checksum pool_guid=0x47641B044A0A279B vdev_path=/dev/sdd2                                          │
│Nov 22 08:34:03 [zed] eid=49 class=checksum pool_guid=0x47641B044A0A279B vdev_path=/dev/sdd2                                          │
│Nov 22 08:34:03 [zed] eid=50 class=zpool pool_guid=0x47641B044A0A279B                                                                 │
│Nov 22 08:34:14 [zed] eid=51 class=checksum pool_guid=0x47641B044A0A279B vdev_path=/dev/sdd2                                          │
└Nov 22 08:34:14 [zed] eid=52 class=checksum pool_guid=0x47641B044A0A279B vdev_path=/dev/sdd2                                          │
─Nov 22 08:34:36 [kernel] [  303.556875] WARNING: can't open objset 54, error 5                                                        │
┌Nov 22 08:34:36 [kernel] [  303.556875] WARNING: can't open objset 54, error 5                                                        │
│Nov 22 08:34:36 [zed] eid=53 class=data pool_guid=0x47641B044A0A279B                                                                  │
│Nov 22 08:34:36 [zed] eid=54 class=checksum pool_guid=0x47641B044A0A279B vdev_path=/dev/sdd2                                          │
│Nov 22 08:34:36 [zed] eid=55 class=checksum pool_guid=0x47641B044A0A279B vdev_path=/dev/sdd2     
```
[  104.333286] WARNING: can't open objset 54, error 5
[  143.578083] WARNING: can't open objset 54, error 5
[  269.774772] WARNING: can't open objset 54, error 5
[  303.556875] WARNING: can't open objset 54, error 5
[  700.581740] WARNING: can't open objset 54, error 5
[ 1105.768923] INFO: task zpool:29976 blocked for more than 122 seconds.
[ 1105.768928]       Tainted: P           OE     5.4.0-1-MANJARO #1
[ 1105.768929] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[ 1105.768932] zpool           D    0 29976   7359 0x00000000
[ 1105.768938] Call Trace:
[ 1105.768953]  ? __schedule+0x2f3/0x770
[ 1105.768962]  schedule+0x43/0xd0
[ 1105.768967]  schedule_preempt_disabled+0x14/0x20
[ 1105.768972]  __mutex_lock.isra.0+0x1ae/0x550
[ 1105.768989]  ? nvlist_xalloc.part.0+0x4a/0xa0 [znvpair]
[ 1105.769136]  spa_all_configs+0x41/0x110 [zfs]
[ 1105.769288]  zfs_ioc_pool_configs+0x17/0x70 [zfs]
[ 1105.769435]  zfsdev_ioctl+0x6e9/0x910 [zfs]
[ 1105.769444]  ? __lru_cache_add+0x75/0xa0
[ 1105.769450]  ? _raw_spin_unlock+0x16/0x30
[ 1105.769457]  ? __handle_mm_fault+0xaf4/0x13a0
[ 1105.769465]  do_vfs_ioctl+0x43d/0x6c0
[ 1105.769472]  ksys_ioctl+0x5e/0x90
[ 1105.769478]  __x64_sys_ioctl+0x16/0x20
[ 1105.769484]  do_syscall_64+0x52/0x180
[ 1105.769491]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[ 1105.769496] RIP: 0033:0x7f29f4e8c25b
[ 1105.769506] Code: Bad RIP value.
[ 1105.769510] RSP: 002b:00007ffe1c640c58 EFLAGS: 00000246 ORIG_RAX: 0000000000000010
[ 1105.769514] RAX: ffffffffffffffda RBX: 000055d72b121470 RCX: 00007f29f4e8c25b
[ 1105.769516] RDX: 00007ffe1c640c80 RSI: 0000000000005a04 RDI: 0000000000000003
[ 1105.769518] RBP: 00007ffe1c644260 R08: 00007f29f51d6010 R09: 0000000000000000
[ 1105.769521] R10: 0000000000000022 R11: 0000000000000246 R12: 000055d72b121470
[ 1105.769523] R13: 0000000000000000 R14: 00007ffe1c640c80 R15: 0000000000000000
[ 1105.769531] INFO: task zfs:30243 blocked for more than 122 seconds.
[ 1105.769534]       Tainted: P           OE     5.4.0-1-MANJARO #1
[ 1105.769535] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[ 1105.769537] zfs             D    0 30243   4723 0x00000000
[ 1105.769542] Call Trace:
[ 1105.769548]  ? __schedule+0x2f3/0x770
[ 1105.769553]  schedule+0x43/0xd0
[ 1105.769557]  schedule_preempt_disabled+0x14/0x20
[ 1105.769562]  __mutex_lock.isra.0+0x1ae/0x550
[ 1105.769575]  ? nvlist_xalloc.part.0+0x4a/0xa0 [znvpair]
[ 1105.769706]  spa_all_configs+0x41/0x110 [zfs]
[ 1105.769851]  zfs_ioc_pool_configs+0x17/0x70 [zfs]
[ 1105.769996]  zfsdev_ioctl+0x6e9/0x910 [zfs]
[ 1105.770003]  ? __lru_cache_add+0x75/0xa0
[ 1105.770009]  ? _raw_spin_unlock+0x16/0x30
[ 1105.770016]  ? __handle_mm_fault+0xaf4/0x13a0
[ 1105.770022]  do_vfs_ioctl+0x43d/0x6c0
[ 1105.770029]  ksys_ioctl+0x5e/0x90
[ 1105.770035]  __x64_sys_ioctl+0x16/0x20
[ 1105.770041]  do_syscall_64+0x52/0x180
[ 1105.770046]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[ 1105.770050] RIP: 0033:0x7f4c21f0f25b
[ 1105.770055] Code: Bad RIP value.
[ 1105.770057] RSP: 002b:00007fff291f0d28 EFLAGS: 00000246 ORIG_RAX: 0000000000000010
[ 1105.770060] RAX: ffffffffffffffda RBX: 000055e0ebc98470 RCX: 00007f4c21f0f25b
[ 1105.770063] RDX: 00007fff291f0d50 RSI: 0000000000005a04 RDI: 0000000000000003
[ 1105.770065] RBP: 00007fff291f4330 R08: 00007f4c220ba010 R09: 0000000000000000
[ 1105.770067] R10: 0000000000000022 R11: 0000000000000246 R12: 000055e0ebc98470
[ 1105.770068] R13: 0000000000000000 R14: 00007fff291f0d50 R15: 0000000000000006
[ 1228.623165] INFO: task zpool:29976 blocked for more than 245 seconds.
[ 1228.623171]       Tainted: P           OE     5.4.0-1-MANJARO #1
[ 1228.623172] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[ 1228.623175] zpool           D    0 29976   7359 0x00000000
[ 1228.623181] Call Trace:
[ 1228.623195]  ? __schedule+0x2f3/0x770
[ 1228.623204]  schedule+0x43/0xd0
[ 1228.623210]  schedule_preempt_disabled+0x14/0x20
[ 1228.623214]  __mutex_lock.isra.0+0x1ae/0x550
[ 1228.623232]  ? nvlist_xalloc.part.0+0x4a/0xa0 [znvpair]
[ 1228.623379]  spa_all_configs+0x41/0x110 [zfs]
[ 1228.623532]  zfs_ioc_pool_configs+0x17/0x70 [zfs]
[ 1228.623680]  zfsdev_ioctl+0x6e9/0x910 [zfs]
[ 1228.623689]  ? __lru_cache_add+0x75/0xa0
[ 1228.623695]  ? _raw_spin_unlock+0x16/0x30
[ 1228.623702]  ? __handle_mm_fault+0xaf4/0x13a0
[ 1228.623710]  do_vfs_ioctl+0x43d/0x6c0
[ 1228.623717]  ksys_ioctl+0x5e/0x90
[ 1228.623723]  __x64_sys_ioctl+0x16/0x20
[ 1228.623729]  do_syscall_64+0x52/0x180
[ 1228.623735]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[ 1228.623741] RIP: 0033:0x7f29f4e8c25b
[ 1228.623751] Code: Bad RIP value.
[ 1228.623754] RSP: 002b:00007ffe1c640c58 EFLAGS: 00000246 ORIG_RAX: 0000000000000010
[ 1228.623759] RAX: ffffffffffffffda RBX: 000055d72b121470 RCX: 00007f29f4e8c25b
[ 1228.623761] RDX: 00007ffe1c640c80 RSI: 0000000000005a04 RDI: 0000000000000003
[ 1228.623763] RBP: 00007ffe1c644260 R08: 00007f29f51d6010 R09: 0000000000000000
[ 1228.623765] R10: 0000000000000022 R11: 0000000000000246 R12: 000055d72b121470
[ 1228.623767] R13: 0000000000000000 R14: 00007ffe1c640c80 R15: 0000000000000000
[ 1228.623776] INFO: task zfs:30243 blocked for more than 245 seconds.
[ 1228.623779]       Tainted: P           OE     5.4.0-1-MANJARO #1
[ 1228.623780] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[ 1228.623782] zfs             D    0 30243   4723 0x00000000
[ 1228.623786] Call Trace:
[ 1228.623792]  ? __schedule+0x2f3/0x770
[ 1228.623797]  schedule+0x43/0xd0
[ 1228.623802]  schedule_preempt_disabled+0x14/0x20
[ 1228.623806]  __mutex_lock.isra.0+0x1ae/0x550
[ 1228.623819]  ? nvlist_xalloc.part.0+0x4a/0xa0 [znvpair]
[ 1228.623953]  spa_all_configs+0x41/0x110 [zfs]
[ 1228.624099]  zfs_ioc_pool_configs+0x17/0x70 [zfs]
[ 1228.624244]  zfsdev_ioctl+0x6e9/0x910 [zfs]
[ 1228.624250]  ? __lru_cache_add+0x75/0xa0
[ 1228.624255]  ? _raw_spin_unlock+0x16/0x30
[ 1228.624262]  ? __handle_mm_fault+0xaf4/0x13a0
[ 1228.624269]  do_vfs_ioctl+0x43d/0x6c0
[ 1228.624276]  ksys_ioctl+0x5e/0x90
[ 1228.624281]  __x64_sys_ioctl+0x16/0x20
[ 1228.624287]  do_syscall_64+0x52/0x180
[ 1228.624293]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[ 1228.624296] RIP: 0033:0x7f4c21f0f25b
[ 1228.624301] Code: Bad RIP value.
[ 1228.624304] RSP: 002b:00007fff291f0d28 EFLAGS: 00000246 ORIG_RAX: 0000000000000010
[ 1228.624307] RAX: ffffffffffffffda RBX: 000055e0ebc98470 RCX: 00007f4c21f0f25b
[ 1228.624309] RDX: 00007fff291f0d50 RSI: 0000000000005a04 RDI: 0000000000000003
[ 1228.624311] RBP: 00007fff291f4330 R08: 00007f4c220ba010 R09: 0000000000000000
[ 1228.624314] R10: 0000000000000022 R11: 0000000000000246 R12: 000055e0ebc98470
[ 1228.624316] R13: 0000000000000000 R14: 00007fff291f0d50 R15: 0000000000000006
[ 1351.516793] INFO: task zpool:29976 blocked for more than 368 seconds.
[ 1351.516801]       Tainted: P           OE     5.4.0-1-MANJARO #1
[ 1351.516803] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[ 1351.516809] zpool           D    0 29976   7359 0x00000000
[ 1351.516821] Call Trace:
[ 1351.516846]  ? __schedule+0x2f3/0x770
[ 1351.516856]  schedule+0x43/0xd0
[ 1351.516863]  schedule_preempt_disabled+0x14/0x20
[ 1351.516869]  __mutex_lock.isra.0+0x1ae/0x550
[ 1351.516889]  ? nvlist_xalloc.part.0+0x4a/0xa0 [znvpair]
[ 1351.517039]  spa_all_configs+0x41/0x110 [zfs]
[ 1351.517196]  zfs_ioc_pool_configs+0x17/0x70 [zfs]
[ 1351.517341]  zfsdev_ioctl+0x6e9/0x910 [zfs]
[ 1351.517349]  ? __lru_cache_add+0x75/0xa0
[ 1351.517356]  ? _raw_spin_unlock+0x16/0x30
[ 1351.517363]  ? __handle_mm_fault+0xaf4/0x13a0
[ 1351.517371]  do_vfs_ioctl+0x43d/0x6c0
[ 1351.517378]  ksys_ioctl+0x5e/0x90
[ 1351.517384]  __x64_sys_ioctl+0x16/0x20
[ 1351.517390]  do_syscall_64+0x52/0x180
[ 1351.517396]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[ 1351.517402] RIP: 0033:0x7f29f4e8c25b
[ 1351.517412] Code: Bad RIP value.
[ 1351.517415] RSP: 002b:00007ffe1c640c58 EFLAGS: 00000246 ORIG_RAX: 0000000000000010
[ 1351.517420] RAX: ffffffffffffffda RBX: 000055d72b121470 RCX: 00007f29f4e8c25b
[ 1351.517422] RDX: 00007ffe1c640c80 RSI: 0000000000005a04 RDI: 0000000000000003
[ 1351.517424] RBP: 00007ffe1c644260 R08: 00007f29f51d6010 R09: 0000000000000000
[ 1351.517427] R10: 0000000000000022 R11: 0000000000000246 R12: 000055d72b121470
[ 1351.517429] R13: 0000000000000000 R14: 00007ffe1c640c80 R15: 0000000000000000
[ 1351.517437] INFO: task zfs:30243 blocked for more than 368 seconds.
[ 1351.517440]       Tainted: P           OE     5.4.0-1-MANJARO #1
[ 1351.517442] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[ 1351.517444] zfs             D    0 30243   4723 0x00000000
[ 1351.517448] Call Trace:
[ 1351.517455]  ? __schedule+0x2f3/0x770
[ 1351.517460]  schedule+0x43/0xd0
[ 1351.517465]  schedule_preempt_disabled+0x14/0x20
[ 1351.517469]  __mutex_lock.isra.0+0x1ae/0x550
[ 1351.517483]  ? nvlist_xalloc.part.0+0x4a/0xa0 [znvpair]
[ 1351.517608]  spa_all_configs+0x41/0x110 [zfs]
[ 1351.517747]  zfs_ioc_pool_configs+0x17/0x70 [zfs]
[ 1351.517885]  zfsdev_ioctl+0x6e9/0x910 [zfs]
[ 1351.517892]  ? __lru_cache_add+0x75/0xa0
[ 1351.517898]  ? _raw_spin_unlock+0x16/0x30
[ 1351.517904]  ? __handle_mm_fault+0xaf4/0x13a0
[ 1351.517910]  do_vfs_ioctl+0x43d/0x6c0
[ 1351.517917]  ksys_ioctl+0x5e/0x90
[ 1351.517923]  __x64_sys_ioctl+0x16/0x20
[ 1351.517929]  do_syscall_64+0x52/0x180
[ 1351.517934]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[ 1351.517938] RIP: 0033:0x7f4c21f0f25b
[ 1351.517943] Code: Bad RIP value.
[ 1351.517946] RSP: 002b:00007fff291f0d28 EFLAGS: 00000246 ORIG_RAX: 0000000000000010
[ 1351.517949] RAX: ffffffffffffffda RBX: 000055e0ebc98470 RCX: 00007f4c21f0f25b
[ 1351.517951] RDX: 00007fff291f0d50 RSI: 0000000000005a04 RDI: 0000000000000003
[ 1351.517953] RBP: 00007fff291f4330 R08: 00007f4c220ba010 R09: 0000000000000000
[ 1351.517955] R10: 0000000000000022 R11: 0000000000000246 R12: 000055e0ebc98470
[ 1351.517957] R13: 0000000000000000 R14: 00007fff291f0d50 R15: 0000000000000006
```

**Comment by inoperable** [comment]
Gonna shift some data around from the suspicious drive and install the new release, I'm not entirely sure now if the error wasn't actually caused by the power state change and the locking bound to happen when waking up a drive. How can I test this to make sure? 

I think I can walk through obvious cases:
- Wake/Sleep cycle in Windows no user-invoked dataset ops
- Wake/Sleep cycle in Windows user-invoked data copy from/to dataset
- Force a Spin-down wait few seconds, make sure drive's in IDLE state, Spin-up and then check if any I/O Errors occur, should be enough, I guess. 

Anything else comes to mind?

**Comment by inoperable** [comment]
Happy to report that I can (since a long time) execute `zpool import -a -f` on Win without instant BSOD ;-) Great Work! Yatta!

**Comment by lundman** [comment]
Create a dataset with casesensitivity=insensitive and try, it is how I use it on Windows.

readonly and driveletter is not something I have look at deeply. :)

**Comment by inoperable** [comment]
Can I then move back to `casesensitivity=mixed` without some side effects? It's not like ZFS will **rename** to **lowercase** right?

**Comment by lundman** [comment]
Isn't it a property that can only be set on creation?

**Comment by inoperable** [comment]
Well, it makes sense, since zpool is imported with `readonly=on` you can't set a property with zfs on it, so I don't know how I can set the property otherwise (from Linux) since this property does not exist on Linux, but there are user properties, so I need (probably) a fully qualified name:property to do so?

**Comment by inoperable** [comment]
Hehehe, yeah, I thought I can change that, but no can't do:
`cannot set property for 'windump': 'casesensitivity' is readonly`

**Comment by lundman** [comment]
The freesid BSOD is fixed with ea2b1a9 : triggered when importing and non-empty unlinked-drain list. I'd be inclined to close this, feel free to open other issues for the additional questions :)
